### PR TITLE
Result processing for RC_SetGlobalProperties

### PR DESCRIPTION
**Pull request by RomanReznichenkoLuxoft**
### Fixes #[11556](https://adc.luxoft.com/jira/browse/FORDTCN-11556)

This PR is **ready** for review.

### Risk
This PR makes **no** API changes.

### Summary
WE have sent a request to RC_SetGlobalProperties on HMI, but we have not processed a response. I added response processing and wrote a test "SUCCESS" for RC_SetGlobalProperties.

### CODE REVIEW

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/reset_global_properties_request.cc [hunk 1]
```diff
@@ -197,12 +208,18 @@ bool ResetGlobalPropertiesRequest::PrepareResponseParameters(
   } else {
     result =
         PrepareResultForMobileResponse(ui_properties_info, tts_properties_info);
+    result = (result || PrepareResultForMobileResponse(ui_properties_info,
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/reset_global_properties_request.cc:211]: @RomanReznichenkoLuxoft I suggest you to create an overload for `PrepareResultForMobileResponse` in the base class which accepts three parameters and use this overload here in one line

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/reset_global_properties_request.cc [hunk 2]
```diff
@@ -197,12 +208,18 @@ bool ResetGlobalPropertiesRequest::PrepareResponseParameters(
...
+                                                       rc_properties_info));
     out_result_code =
         PrepareResultCodeForResponse(ui_properties_info, tts_properties_info);
-    out_response_info = app_mngr::commands::MergeInfos(tts_properties_info,
-                                                       tts_response_info_,
-                                                       ui_properties_info,
-                                                       ui_response_info_);
+    if (rc_result_ != hmi_apis::Common_Result::INVALID_ENUM) {
+      out_result_code = std::max(
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/reset_global_properties_request.cc:216]: @RomanReznichenkoLuxoft same as above - please create overload for `PrepareResultCodeForResponse` in the base class which accepts three arguments and use it (note that `MergeInfos()` already has overloaded function which accepts three arguments)

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/test/commands/mobile/reset_global_properties_test.cc [hunk 3]
```diff
@@ -282,6 +282,42 @@ TEST_F(ResetGlobalPropertiesRequestTest,
...
+
+  command_->Run();
+
+  event.set_smart_object(*msg_);
+  EXPECT_CALL(mock_rpc_service_,
+              ManageMobileCommand(
+                  MobileResultCodeIs(mobile_apis::Result::eType::SUCCESS),
+                  am::commands::Command::SOURCE_SDL));
+
+  command_->on_event(event);
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/test/commands/mobile/reset_global_properties_test.cc:318]: @RomanReznichenkoLuxoft you can actually move all the code related to this event to one place:
```suggestion
  (*msg_)[am::strings::params][am::hmi_response::code] = hmi_apis::Common_Result::SUCCESS;  
  
  Event event(hmi_apis::FunctionID::RC_SetGlobalProperties);
  event.set_smart_object(*msg_);
  command_->on_event(event);
```

**Review by VadymLuchko — APPROVED** [review]

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/reset_global_properties_request.cc (hunk 2 above)
Comment (RomanReznichenkoLuxoft) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/reset_global_properties_request.cc:214]: Done

File: src/components/application_manager/src/commands/command_request_impl.cc [hunk 4]
```diff
@@ -748,6 +748,18 @@ bool CommandRequestImpl::PrepareResultForMobileResponse(
...
 
+bool CommandRequestImpl::PrepareResultForMobileResponse(
+    ResponseInfo& out_first,
+    ResponseInfo& out_second,
+    ResponseInfo& out_third) const {
+  SDL_LOG_AUTO_TRACE();
+  bool result = CheckResult(out_first, out_second) ||
+                CheckResult(out_second, out_first) ||
+                CheckResult(out_third, out_second) ||
+                CheckResult(out_second, out_third);
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/src/commands/command_request_impl.cc:759]: @RomanReznichenkoLuxoft as this is OR it is enough to write like this:
```suggestion
  bool result = CheckResult(out_first, out_second) ||
                CheckResult(out_second, out_third);
```

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/test/commands/mobile/reset_global_properties_test.cc (hunk 3 above)
Comment (RomanReznichenkoLuxoft) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/test/commands/mobile/reset_global_properties_test.cc:318]: done

File: src/components/application_manager/src/commands/command_request_impl.cc (hunk 4 above)
Comment (RomanReznichenkoLuxoft) [reply src/components/application_manager/src/commands/command_request_impl.cc:759]: done

File: src/components/application_manager/include/application_manager/commands/command_request_impl.h [hunk 5]
```diff
@@ -273,6 +273,22 @@ class CommandRequestImpl : public CommandImpl,
...
                                       ResponseInfo& out_second) const;
 
+  /**
+   * @brief Checks result code from HMI for splitted RPC
+   * and returns parameter for sending to mobile app.
+   * @param first contains result_code from HMI response and
+   * interface that returns response
+   * @param second contains result_code from HMI response and
+   * interface that returns response
+   * * @param third contains result_code from HMI response and
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/include/application_manager/commands/command_request_impl.h:283]:
```suggestion
   * @param third contains result_code from HMI response and
```

File: src/components/application_manager/src/commands/command_request_impl.cc [hunk 6]
```diff
@@ -787,6 +798,43 @@ mobile_apis::Result::eType CommandRequestImpl::PrepareResultCodeForResponse(
...
+    first_result = first.result_code;
+  }
+  if (!second.is_unsupported_resource) {
+    second_result = second.result_code;
+  }
+  if (!third.is_unsupported_resource) {
+    third_result = third.result_code;
+  }
+  mobile_apis::Result::eType result_code = MessageHelper::HMIToMobileResult(
+      std::max(std::max(first_result, second_result), third_result));
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/src/commands/command_request_impl.cc:834]: @RomanReznichenkoLuxoft I think you can use `std::max` version which accepts initializer list like:
```suggestion
      std::max({first_result, second_result, third_result});
```

File: src/components/application_manager/src/commands/command_request_impl.cc [hunk 7]
```diff
@@ -787,6 +798,43 @@ mobile_apis::Result::eType CommandRequestImpl::PrepareResultCodeForResponse(
...
+  return result_code;
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/src/commands/command_request_impl.cc:835]: @RomanReznichenkoLuxoft return and previous line of code can be merged. No sense to create a `result_code` variable if you just returning it

File: src/components/application_manager/src/commands/command_request_impl.cc [hunk 8]
```diff
@@ -787,6 +798,43 @@ mobile_apis::Result::eType CommandRequestImpl::PrepareResultCodeForResponse(
   return result_code;
 }
 
+mobile_apis::Result::eType CommandRequestImpl::PrepareResultCodeForResponse(
+    const ResponseInfo& first,
+    const ResponseInfo& second,
+    const ResponseInfo& third) {
+  SDL_LOG_AUTO_TRACE();
+  if (IsResultCodeUnsupported(first, second) ||
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/src/commands/command_request_impl.cc:806]: @RomanReznichenkoLuxoft I believe all that function can be replaced like
```
const auto first_comparison = PrepareResultCodeForResponse(first, second);
const auto second_comparison = PrepareResultCodeForResponse(second, third);
const auto third_comparison = PrepareResultCodeForResponse(first, third);

return std::max({first_comparison, second_comparison, third_comparison});
```

**Review by AKalinich-Luxoft — APPROVED** [review]

**Review by ychernysheva — APPROVED** [review]

**Pull request closed without merging**